Improve and restructure LS documentation (#36717)
ballerina-platform/ballerina-lang: docs/language-server/ExtensionDevelopmentBestPractices.md

# Extension Development Best Practices

## What is Language Server Extension

The language server protocol contains various language features and the ballerina's LSP implementation exposes a set of
extension points where third parties can write their own extensions for these language features and enrich the user
experience. At the moment following extension points are provided.

1. [Extended Language Server Services](WritingExtendedServices.md)
2. [Compiler Plugin Code Actions](CompilerPluginCodeActions.md)
3. Code Action Extension (Java SPI)
4. Completion Extension (Java SPI)

> 5. ~~Formattin Extenstion (Java SPI)~~

Reviewer: ```suggestion
5. ~~Formatting Extension (Java SPI)~~
```

6. ~~Diagnostic Extension (Java SPI)~~

__NOTE__: Formatting and Diagnostic extensions are not used currently.

In order to compose an extension, the developer should have an understanding of the underlying supporting APIs which

> being used heavily in the existing implementations. Following are the main APIs being used in the implementation.

Reviewer: ```suggestion
are being used heavily in the existing implementations. Following are the main APIs being used in the implementation.
```

- Hence **Don't Use the API**

## Important tips for various Usages of APIs

**1. How do we check for the node is of a specific kind**  
**Answer**: _Use the `SyntaxKind` to check and avoid the use of **InstanceOf** checks_

**2.How to check a symbol is of a specific kind of Symbol**  
**Answer**: _Use the `SymbolKind` and `TypeKind` for the check_

**3. How to get the children of a Syntax node**  
**Answer:** _Use the APIs provided in each of the node. For example the `FunctionDefinition` contains an API
called `qualifierList` to get the qualifiers in the function. Try to avoid the usage of `children` API since it is too
generic and includes whitesapces as well_

**4. How to access the position details**  
**Answer:** _Use either the offset or line-column approach. In either case, stick to one usage without mixing them both.
Better option would be to use the offset approach_

**5. How to Access the Source code from a syntax node**  
**Answer:** _Use the `toSourceCode` API. **NEVER** use the `toString` API_

**6. How to traverse the Syntax Tree**  

> **Answer:** _Best and recomended optionis to use a node visitor. The `NodeVisitor` facilitates this. Each `visit` method

Reviewer: ```suggestion
**Answer:** _Best and recomended option is to use a node visitor. The `NodeVisitor` facilitates this. Each `visit` method
```
